File: src/shapes/codex.ts
import { createReadStream } from "node:fs";
import { createInterface } from "node:readline";
import { basename } from "node:path";
import { asObj, str, truncate, uidOf, type ParsedEvent, type ParsedFile, type Parser } from "../types.js";

/**
 * Codex CLI rollouts — ~/.codex/sessions/**\/rollout-*.jsonl
 *
 * Structurally unlike Claude: the file is a stream of MIXED record types and only
 * `response_item` carries transcript. `event_msg` and `token_usage_record` are UI
 * bookkeeping and are counted but never indexed — indexing them floods search with
 * noise that looks like conversation.
 */
export const parseCodex: Parser = async (filePath) => {
  const fileKey = basename(filePath);
  const events: ParsedEvent[] = [];
  const typeCounts: Record<string, number> = {};
  let lines = 0, badLines = 0, seq = 0;
  let cwd: string | null = null, model: string | null = null;
  // Fallback identity: a rollout truncated before its session_meta line still needs a stable id.
  let sessionUuid = fileKey.replace(/^rollout-/, "").replace(/\.jsonl$/, "");
  let startedAt: string | null = null, endedAt: string | null = null;
  let description: string | null = null;

  const push = (role: string, text: string, ts: string | null, partIdx: number) => {
    const t = text.trim();
    if (!t) return;
    if (!description && role === "user") description = truncate(t, 200);
    events.push({ uid: uidOf("codex", fileKey, seq * 1000 + partIdx), seq, role, ts, text: truncate(t) });
  };

  const rl = createInterface({ input: createReadStream(filePath, "utf8") });
  for await (const line of rl) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    lines++; seq++;

    let parsed: unknown;
    try { parsed = JSON.parse(trimmed); } catch { badLines++; continue; }
    const rec = asObj(parsed);
    if (!rec) { badLines++; continue; }

    const type = str(rec.type) ?? "unknown";
    typeCounts[type] = (typeCounts[type] ?? 0) + 1;
    const ts = str(rec.timestamp);
    if (ts) { if (!startedAt) startedAt = ts; endedAt = ts; }

    const payload = asObj(rec.payload);
    if (!payload) continue;

    if (type === "session_meta") {
      sessionUuid = str(payload.id) ?? str(payload.session_id) ?? sessionUuid;
      cwd = str(payload.cwd) ?? cwd;
      continue;
    }
    if (type === "turn_context") {
      model = model ?? str(payload.model);
      cwd = cwd ?? str(payload.cwd);
      continue;
    }
    if (type !== "response_item") continue;   // event_msg / token_usage_record are UI noise

    const ptype = str(payload.type);
    if (ptype === "message" || ptype === "agent_message") {
      const role = ptype === "agent_message" ? "assistant" : (str(payload.role) ?? "unknown");
      if (Array.isArray(payload.content)) {
        let i = 0;
        for (const raw of payload.content) {
          const item = asObj(raw);
          const t = item ? (str(item.text) ?? "") : "";
          if (t) push(role, t, ts, i);
          i++;
        }
      } else {
        push(role, str(payload.message) ?? str(payload.text) ?? "", ts, 0);
      }
    } else if (ptype === "reasoning") {
      // Codex emits reasoning as its own record type. Kept as a distinct role so it can
      // be searched — or excluded — separately. Claude's shape has no equivalent.
      const parts: string[] = [];
      if (Array.isArray(payload.summary)) {
        for (const raw of payload.summary) {
          const s = asObj(raw);
          const t = s ? (str(s.text) ?? "") : "";
          if (t) parts.push(t);
        }
      }
      push("reasoning", parts.join("\n"), ts, 0);
    } else if (ptype === "custom_tool_call" || ptype === "function_call") {
      const name = str(payload.name) ?? "tool";
      const input = str(payload.input) ?? str(payload.arguments) ?? "";
      push("tool_use", `[tool_call ${name}] ${truncate(input, 2000)}`, ts, 0);
    } else if (ptype === "custom_tool_call_output" || ptype === "function_call_output") {
      push("tool_result", `[tool_output] ${truncate(str(payload.output) ?? "", 4000)}`, ts, 0);
    }
  }

  // Codex rollouts carry no title record — the caller falls back to description.
  return { sessionUuid, cwd, model, events, lines, badLines, typeCounts, startedAt, endedAt, description, title: null, gitBranch: null };
};

File: python/relicpy/shapes/codex.py
"""Codex CLI rollouts — ~/.codex/sessions/YYYY/MM/DD/rollout-<ts>-<uuid>.jsonl.

A DIFFERENT SHAPE, not a rename. Only `response_item` records are transcript;
`event_msg` and `token_usage_record` are UI noise and are counted, never indexed.
"""

from __future__ import annotations

import json
import os
import re

from ..models import ParsedEvent, ParsedFile
from ..types import as_obj, s, truncate, uid_of


def parse(file_path: str) -> ParsedFile:
    file_key = os.path.basename(file_path)
    events: list[ParsedEvent] = []
    lines = bad_lines = seq = 0
    cwd = model = None
    # Fallback identity: a rollout truncated before its session_meta line still needs
    # a stable id.
    session_uuid = re.sub(r"\.jsonl$", "", re.sub(r"^rollout-", "", file_key))
    started_at = ended_at = description = None

    def push(role: str, text: str, ts, part_idx: int) -> None:
        nonlocal description
        t = (text or "").strip()
        if not t:
            return
        if not description and role == "user":
            description = truncate(t, 200)
        # seq * 1000 + partIdx: one line can carry several content parts, and each
        # needs its own uid. The multiplier is the ceiling on parts per line.
        events.append(ParsedEvent(uid=uid_of("codex", file_key, seq * 1000 + part_idx),
                                  seq=seq, role=role, ts=ts, text=truncate(t)))

    with open(file_path, "r", encoding="utf-8", errors="replace") as fh:
        for line in fh:
            trimmed = line.strip()
            if not trimmed:
                continue
            lines += 1
            seq += 1
            try:
                rec = as_obj(json.loads(trimmed))
            except Exception:
                bad_lines += 1
                continue
            if rec is None:
                bad_lines += 1
                continue

            typ = s(rec.get("type")) or "unknown"
            ts = s(rec.get("timestamp"))
            if ts:
                if not started_at:
                    started_at = ts
                ended_at = ts

            payload = as_obj(rec.get("payload"))
            if not payload:
                continue

            if typ == "session_meta":
                session_uuid = s(payload.get("id")) or s(payload.get("session_id")) or session_uuid
                cwd = s(payload.get("cwd")) or cwd
                continue
            if typ == "turn_context":
                model = model or s(payload.get("model"))
                cwd = cwd or s(payload.get("cwd"))
                continue
            if typ != "response_item":
                continue

            ptype = s(payload.get("type"))
            if ptype in ("message", "agent_message"):
                role = "assistant" if ptype == "agent_message" else (s(payload.get("role")) or "unknown")
                content = payload.get("content")
                if isinstance(content, list):
                    for i, raw in enumerate(content):
                        item = as_obj(raw)
                        t = (s(item.get("text")) or "") if item else ""
                        if t:
                            push(role, t, ts, i)
                else:
                    push(role, s(payload.get("message")) or s(payload.get("text")) or "", ts, 0)
            elif ptype == "reasoning":
                # Codex emits reasoning as its own record type. Kept as a distinct role
                # so it can be searched — or excluded — separately.
                parts = []
                summary = payload.get("summary")
                if isinstance(summary, list):
                    for raw in summary:
                        item = as_obj(raw)
                        t = (s(item.get("text")) or "") if item else ""
                        if t:
                            parts.append(t)
                push("reasoning", "\n".join(parts), ts, 0)
            elif ptype in ("custom_tool_call", "function_call"):
                name = s(payload.get("name")) or "tool"
                inp = s(payload.get("input")) or s(payload.get("arguments")) or ""
                push("tool_use", f"[tool_call {name}] {truncate(inp, 2000)}", ts, 0)
            elif ptype in ("custom_tool_call_output", "function_call_output"):
                push("tool_result", f"[tool_output] {truncate(s(payload.get('output')) or '', 4000)}", ts, 0)

    # Codex rollouts carry no title record — the caller falls back to description.
    return ParsedFile(session_uuid=session_uuid, cwd=cwd, model=model, events=events,
                      lines=lines, bad_lines=bad_lines, started_at=started_at,
                      ended_at=ended_at, description=description, title=None, git_branch=None)
